src/WebJobs.Extensions.Http/ClaimsPrincipalBindingProvider.cs

```diff
@@ -0,0 +1,121 @@
+    {
+        private static readonly Task<IBinding> NullBinding = Task.FromResult<IBinding>(null);
+        private bool _isEasyAuthEnabled;
+
+        public ClaimsPrincipalBindingProvider(INameResolver nameResolver)
+        {
+            _isEasyAuthEnabled = ClaimsIdentityHelper.IsEasyAuthEnabled(nameResolver);
+        }
+
+        public Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            if (context.Parameter.ParameterType != typeof(ClaimsPrincipal))
+            {
+                return NullBinding;
+            }
+
+            return Task.FromResult<IBinding>(new ClaimsIdentityBinding(context.Parameter, _isEasyAuthEnabled));
+        }
+
+        private class ClaimsIdentityBinding : IBinding
+        {
+            private readonly ParameterInfo _parameter;
+            private readonly bool _isEasyAuthEnabled;
+
+            public ClaimsIdentityBinding(ParameterInfo parameter, bool isEasyAuthEnabled)
+            {
+                _parameter = parameter;
+                _isEasyAuthEnabled = isEasyAuthEnabled;
+            }
+
+            public bool FromAttribute
+            {
+                get { return false; }
+            }
+
+            public Task<IValueProvider> BindAsync(BindingContext context)
+            {
+                if (context == null)
+                {
+                    throw new ArgumentNullException("context");
+                }
+
+                IReadOnlyDictionary<string, object> bindingData = context.BindingData;
+                ClaimsPrincipal principal = GetClaimsPrincipalFromBindingData(bindingData);
+
+                var valueProvider = new SimpleValueProvider(typeof(ClaimsPrincipal), principal, principal?.Identity?.Name);
+                return Task.FromResult<IValueProvider>(valueProvider);
+            }
+
+            private ClaimsPrincipal GetClaimsPrincipalFromBindingData(IReadOnlyDictionary<string, object> bindingData)
+            {
+                object headers;
+                if (bindingData.TryGetValue(HttpTriggerAttributeBindingProvider.HttpHeadersKey, out headers))
+                {
+                    var identities = new List<ClaimsIdentity>(ClaimsIdentityHelper.GetClaimsIdentitiesFromHeaders(headers as IDictionary<string, string>, _isEasyAuthEnabled)); 
+
+                    if (identities.Count == 0)
+                    {
+                        // According to the source https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/security/principal/genericidentity.cs
+                        // GenericIdentity is unauthenticated if and only if the name is set to the empty string
+                        var unauthentiatedIdentity = new GenericIdentity(name: string.Empty);
+                        identities.Add(unauthentiatedIdentity);
```

Comment:
If there are no identities, why do we create an identity at all? Why not have a ClaimsPrincipal with no identities?

Reply:
so if we don't create an identity, then the user can't do `claimsPrincipal.Identity.IsAuthenticated`, because `claimsPrincipal.Identity` is null. You could argue whether that is required, but either way, I should probably populate it the same here as I do at line 93.

---
In reply to: [216399689](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/404#discussion_r216399689) [](ancestors = 216399689)